modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryEventBuffer.java

```diff
@@ -463,26 +533,47 @@ private CacheContinuousQueryEntry filteredEntry(long cntr, long filtered) {
 
                     lastProc = pos;
 
-                    if (pos == entries.length - 1) {
-                        Arrays.fill(entries, null);
+                    if (pos == entries.length - 1)
+                        rollOver(startCntr + BUF_SIZE, filtered, entry.topologyVersion());
+                }
+                else if (endCntr < ackedUpdCntr0)
+                    rollOver(ackedUpdCntr0 + 1, 0, entry.topologyVersion());
 
-                        Batch nextBatch = new Batch(this.startCntr + BUF_SIZE,
-                            filtered,
-                            entries,
-                            entry.topologyVersion());
+                return res;
+            }
+        }
 
-                        entries = null;
+        /**
+         * @param topVer Topology version of current processing entry.
+         */
+        private synchronized void tryRollOver(AffinityTopologyVersion topVer) {
+            if (entries == null)
+                return;
 
-                        assert curBatch.get() == this;
+            long ackedUpdCntr0 = ackedUpdCntr;
 
-                        curBatch.set(nextBatch);
-                    }
-                }
-                else
-                    return res;
-            }
+            if (endCntr < ackedUpdCntr0)
```

Comment:
You should also check `this` against `curBatch`, since `curBatch` can be changed concurrently and in this case you will fail with assert on `rollOver`

Reply:
I think it's enough to check `entries == null` under `this` monitor synchronized block, this will guarantee us that `batch` will not be changed concurrently.